File: src/main/java/qmech/helper/tileentity/CustomModelBase.java
package qmech.helper.tileentity;

import net.minecraft.client.model.ModelBase;
import net.minecraft.client.model.ModelRenderer;
import net.minecraft.entity.Entity;
import net.minecraft.entity.player.EntityPlayer;

/**
 * Created by anshuman on 14-05-2014.
 */
public abstract class CustomModelBase extends ModelBase {

    public CustomModelBase() {
        initTextureSize();
        initShapes();
    }

    public void render(Entity player, float f, float f1, float f2, float f3, float f4, float f5)
    {
        super.render(player, f, f1, f2, f3, f4, f5);
        setRotationAngles(f, f1, f2, f3, f4, f5, player);
    }

    public abstract void renderShapes(Entity player, float f, float f1, float f2, float f3, float f4, float f5);

    public abstract void initTextureSize();

    public abstract void initShapes();

    public void setRotation(ModelRenderer model, float x, float y, float z)
    {
        model.rotateAngleX = x;
        model.rotateAngleY = y;
        model.rotateAngleZ = z;
    }

    public void setRotationAngles(float f, float f1, float f2, float f3, float f4, float f5, EntityPlayer player)
    {
        super.setRotationAngles(f, f1, f2, f3, f4, f5, player);
    }

}

File: src/main/scala/geek/lawsof/physics/lib/objects/block/render/model/CustomModelBase.scala
package geek.lawsof.physics.lib.objects.block.render.model

import net.minecraft.client.model.{TextureOffset, ModelRenderer, ModelBase}
import net.minecraft.entity.Entity

/**
 * Created by anshuman on 28-05-2014.
 */
abstract class CustomModelBase extends ModelBase {
  initShapes

  def texSize: (Int, Int)

  override def render(p: Entity, f1: Float, f2: Float, f3: Float, f4: Float, f5: Float, f6: Float) = {
    super.render(p, f1, f2, f3, f4, f5, f6)
    setRotationAngles(f1, f2, f3, f4, f5, f6, p)
    renderShapes(p, f1, f2, f3, f4, f5, f6)
  }

  def renderShapes(p: Entity, f1: Float, f2: Float, f3: Float, f4: Float, f5: Float, f6: Float)

  def initShapes

  def setRotationForModel(model: ModelRenderer, f1: Int, f2: Int, f3: Int) = {
    model.rotateAngleX = f1
    model.rotateAngleY = f2
    model.rotateAngleZ = f3
  }

  def initCubeShape(offset: TextureOffset = new TextureOffset(0, 0),
                    corner: (Int, Int, Int) = (0, 0, 0),
                    size: (Int, Int, Int) = (16, 16, 16),
                    rotationPoint: (Float, Float, Float) = (0F, 0F, 0F),
                    rotation: (Int, Int, Int) = (0, 0, 0),
                    mirror: Boolean = false) = {
    var block: ModelRenderer = new ModelRenderer(this, offset.textureOffsetX, offset.textureOffsetY)
    block.addBox(corner._1, corner._2, corner._3, size._1, size._2, size._3)
    block.setRotationPoint(rotationPoint._1, rotationPoint._2, rotationPoint._3)
    block.setTextureSize(texSize._1, texSize._2)
    setRotationForModel(block, rotation._1, rotation._2, rotation._3)
    block.mirror = mirror
  }

  def renderShape(block: ModelRenderer, f6: Float) = block.render(f6)
}
